# DATA IMPORT & CLEANING

In [1]:
#Importing the libraries
import pandas as pd
import os
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import math

os.chdir("") #your directory
os.getcwd()

'/Users/jasonrayen/Downloads/Jason Masters/Sem 3/AIT 664 (Information, Visualization & Representation)/Project'

In [2]:
#Importing the dataset
dataset_s1 = pd.read_excel("S1 File.xlsx")
dataset_s2 = pd.read_excel("S2 File.xlsx")

In [3]:
#Creating a copy of the dataframe
dataset_s1_cleaned = dataset_s1

In [4]:
#Number of rows and columns in the dataset - S1
print(dataset_s1_cleaned.shape)

(801, 79)


In [5]:
#Number of rows and columns in the dataset - S2
print(dataset_s1_cleaned.shape)

(801, 79)


In [6]:
#Checking the number of Null values in the dataset
dataset_s1_cleaned.isnull().sum()

Census tract                                      0
The number of crime incidents                     0
Total population                                  0
Male population                                   0
Percent of male population                        3
                                                 ..
Percent of population of public administration    4
Population of other services                      0
Percent of population of other services           4
Median income                                     5
Mean income                                       3
Length: 79, dtype: int64

In [7]:
# Assuming df is your DataFrame
null_counts = dataset_s1_cleaned.isnull().sum()
variables_with_nulls = null_counts[null_counts > 0]

print(f"Number of variables with null values: {len(variables_with_nulls)}")
print(variables_with_nulls)

Number of variables with null values: 40
Percent of male population                                                                                      3
Percent of female population                                                                                    3
Median age                                                                                                      3
Percent of under 16 years                                                                                       3
Percent of over 65 years                                                                                        3
Percent of occupied housing units                                                                               3
Percent of vacant housing units                                                                                 3
Median housing value                                                                                            9
Percent of less than 9th grade                 

In [8]:
#Total Null values
dataset_s1_cleaned.isnull().sum().sum()

141

In [9]:
#Columns that have null values
columns_with_null = dataset_s1_cleaned.columns[dataset_s1_cleaned.isnull().any()].tolist()
print("Columns with null values:", columns_with_null)

Columns with null values: ['Percent of male population', 'Percent of female population', 'Median age', 'Percent of under 16 years', 'Percent of over 65 years', 'Percent of occupied housing units', 'Percent of vacant housing units', 'Median housing value', 'Percent of less than 9th grade', 'Percent of 9th to 12th grade', 'Percent of high school graduate', 'Percent of some college', 'Percent of associate’s degree', 'Percent of bachelor’s degree', 'Percent of graduate or professional degree', 'Percent of high school graduate or higher', 'Percent of bachelor’s degree of higher', 'Percent of less than high school graduate', 'Percent of less than bachelor’s degree', 'Percent of over 16 years', 'Percent of population in labor force', 'Percent of population not in labor force', 'Percent of employed population', 'Percent of unemployed population', 'Percent of population not in labor force and unemployed population', 'Percent of population of agriculture, forestry, fishing, hunting, and mining',

In [10]:
#Number of Unique Census tracts in the dataset
print("Number of unique census values is: {}".format(dataset_s1_cleaned["Census tract"].nunique()))

Number of unique census values is: 801


We have to find out the crime incidents occuring per census. So let's consider the columns which are really necessary for the analysis. Before that, Let's check the correlation matrix and remove the correlated variables.

# CORRELATION MATRIX 

In [11]:
#Correlation values
correlation_matrix = dataset_s1_cleaned.corr()
print("Correlation Matrix:")
print(correlation_matrix)

Correlation Matrix:
                                                Census tract  \
Census tract                                        1.000000   
Total population                                   -0.104293   
Male population                                    -0.115336   
Percent of male population                         -0.176175   
Female population                                  -0.087792   
...                                                      ...   
Percent of population of public administration     -0.031653   
Population of other services                        0.051515   
Percent of population of other services             0.254840   
Median income                                      -0.317694   
Mean income                                        -0.354512   

                                                Total population  \
Census tract                                           -0.104293   
Total population                                        1.000000   
Male po

In [12]:
#Correlation values - greater than 0.9
high_correlation_pairs = []
for i in range(len(correlation_matrix.columns)):
    for j in range(i+1, len(correlation_matrix.columns)):
        if abs(correlation_matrix.iloc[i, j]) > 0.9:
            high_correlation_pairs.append((correlation_matrix.columns[i], correlation_matrix.columns[j], correlation_matrix.iloc[i, j]))


for pair in high_correlation_pairs:
    print(pair[0], "-", pair[1], ": correlation =", pair[2])

Total population - Male population : correlation = 0.9760193250538579
Total population - Female population : correlation = 0.974816673759532
Total population - Over 25 years : correlation = 0.956184465580599
Total population - Over 16 years : correlation = 0.9797757158651133
Total population - Population in labor force : correlation = 0.9083976773372419
Male population - Female population : correlation = 0.9028947188138949
Male population - Over 25 years : correlation = 0.9343491448970911
Male population - Over 16 years : correlation = 0.9626040652492975
Percent of male population - Percent of female population : correlation = -1.0
Female population - Over 25 years : correlation = 0.9309831609766871
Female population - Over 16 years : correlation = 0.9486230065177955
Percent of under 16 years - Percent of over 16 years : correlation = -1.0
Total housing units - Occupied housing units : correlation = 0.9873782587114923
Total housing units - Over 25 years : correlation = 0.90302779760379

In [13]:
#Removing certain variables with high correlation of 0.9 and above
columns_to_remove = ['Male population', 'Female population', 'Over 25 years', 'Over 16 years', 'Population in labor force'
                    , 'Occupied housing units', 'Percent of bachelor’s degree of higher', 'Percent of less than bachelor’s degree'
                    , 'Percent of high school graduate or higher', 'Percent of less than high school graduate', 'Percent of population in labor force', 
                     'Population not in labor force', 'Population not in labor force and unemployed population', 'Median income', 'Employed population',
                    'Percent of male population', 'Percent of female population', 'Percent of under 16 years', 'Percent of over 16 years', 
                    'Percent of occupied housing units', 'Percent of vacant housing units', 'Percent of population not in labor force', 
                     'Percent of population not in labor force and unemployed population', 'Employed population', 'Unemployed population',
                    'Percent of over 65 years', 'Under 16 years', 'Over 65 years', 'Population of agriculture, forestry, fishing, hunting, and mining',
                    'Population of construction', 'Population of manufacturing', 'Population of wholesale trade', 'Population of retail trade',
                    'Population of transportation, warehousing, and utilities', 'Population of information', 'Population of finance, insurance, real estate, rental, and leasing',
                    'Population of professional, scientific, management, administrative, and waste management services', 'Population of educational services, health care, and social assistance',
                    'Population of arts, entertainment, recreation, accommodation, and food services', 'Population of public administration', 'Population of other services']
dataset_s1_cleaned = dataset_s1_cleaned.drop(columns = columns_to_remove)

In [14]:
#Now again checking the Null values - Post correlation removal
dataset_s1_cleaned.isnull().sum()

Census tract                                                                                                    0
The number of crime incidents                                                                                   0
Total population                                                                                                0
Median age                                                                                                      3
White                                                                                                           0
Black or African American                                                                                       0
American Indian or Alaska Native                                                                                0
Asian                                                                                                           0
Native Hawaiian and Other Pacific Islander                                              

In [15]:
#Filling up the null values with the mean of the column
for column in dataset_s1_cleaned.columns:
    if dataset_s1_cleaned[column].isnull().any():
        mean_value = dataset_s1_cleaned[column].mean()
        dataset_s1_cleaned[column] = dataset_s1_cleaned[column].fillna(mean_value)
        dataset_s1_cleaned[column] = dataset_s1_cleaned[column].round(2)

In [16]:
#Final Check of Null values
dataset_s1_cleaned.isnull().sum()

Census tract                                                                                                    0
The number of crime incidents                                                                                   0
Total population                                                                                                0
Median age                                                                                                      0
White                                                                                                           0
Black or African American                                                                                       0
American Indian or Alaska Native                                                                                0
Asian                                                                                                           0
Native Hawaiian and Other Pacific Islander                                              

We have the same 801 values but now with no null values. All null values have been fixed with the help of correlation and the mean of the columns are taken and filled in the missing spaces

In [17]:
#Transfering output file to R
dataset_s1_cleaned.to_csv(r'dataset_s1_cleaned.csv', index=False)